### PR TITLE
Change name of data load functions

### LINKED ISSUE #114: Add `load_data` file to `pipeline.utils`
```python
def load_data(file):
    """Load PE and injection file created from `gwinferno.preprocess.data_collection.load_posterior_samples_and_injections()`.

    Args:
        file (str): Path to PE and Injection file

    Returns:
        pedict (dict): dictionary of PE samples
        injdict (dict): dictionary of injection data
        constants (dict): dictionary of constants like total generated injections
        param_names (list of str): list of parameter names
    """
    data = az.from_netcdf(file)
    print(f'data file {file} loaded')

    pedict = {k: jnp.asarray(data.pe_data.posteriors.sel(param=k).values) for k in data.pe_data.param.values}
    injdict = {k: jnp.asarray(data.inj_data.injections.sel(param=k).values) for k in data.inj_data.param.values}

    param_names = list(data.pe_data.param.values)
    
    total_inj = data.inj_data.attrs['total_generated']
    obs_time = data.inj_data.attrs['analysis_time']
    nObs = data.pe_data.posteriors.shape[0]

    constants = {
        'total_inj': total_inj, 
        'obs_time': obs_time, 
        'nObs': nObs
        }

    return pedict, injdict, constants, param_names
```

### PR DESCRIPTION
**Pull request by jaxengodfrey**
I changed the name of `data_collection.load_posterior_samples_and_injections` to `data_collection.save_posterior_samples_and_injectioons` because I added a function in `pipeline.utils` called `load_pe_and_injections_as_dict` and wanted to avoid confusion for which function to use in an analysis script. 

@bfarr does this convention make sense to you or do you have any ideas of how to make the function uses clear?

Closes #114

### CODE REVIEW

**Comment by bfarr** [comment]
There's a _lot_ of data manipulation, and it's a little hard to follow. A few thoughts:

### Separating input and output

From reading through the code it seems like the intention is for the user to (rarely) use the routines in `preprocess/data_collection.py` to turn an event-name-to-metadata mapping (including paths to sample-containing HDF5 files) into a single `az.InferenceData` object that's saved to file. That file is later read in for analyses, and the routine for reading it and returning the dictionaries we typically use during inference is in `pipeline/utils.py`. Is that right? Does the library just use `az.InferenceData` objects because they're convenient for reading/writing, or do we use them elsewhere? Should we consider using them more extensively (e.g., providing them to the likelihood instead of dictionaries)?

It's generally good practice to have dedicated functions for reading and writing; e.g., [`save_posterior_samples_and_injections()` in `gwinferno/preprocess/data_collection.py`](https://github.com/FarrOutLab/GWInferno/blob/9dfdd57c955fbea2b0e65f2f36602c970c117735/gwinferno/preprocess/data_collection.py#L194) both uses a `json` file to find and load PE samples, construct an `InferenceData` object, and save that to file with a prescriptive file path.

### More descriptive function names

Separate from the choices we make above, it would be helpful to have more descriptive funcion names. Taking `data_collection.py` as an example, if I'm understanding things properly I think the following renames and modifications would make the code easier to follow:

- `run_map` > `catalog_metadata`
  
- `collect_data(run_map)` >`catalog_dict_from_metadata(catalog_metadata)`
  
- `format_data(posteriors)` > `catalog_xarray_from_dict(catalog_dict)`
  
- `evaluate_prior(full_catalog, param_names)` > `append_prior_to_catalog(catalog_xarray, param_names)`
  
- `save_posterior_samples_and_injections(key_file, injfile, parameter_names, outdir...)` > `inf_data_from_metadata_and_injections(catalog_metadata, injfile, parameter_names, ...)` where it returns the inference data object but doesn't write anthing to file.
  
- Then the user calls the `.to_netcdf()` method on the returned object, or we create a minimal wrapper function to do it, e.g., `save_catalog_inf_data(filename, catalog_inf_data)`. I think the only reason to have a function that forces a formatting to the filename is if we intend to have meaningful metadata as part of the filename, which doesn't seem necessary.

**Comment by bfarr** [comment]
There's a _lot_ of data manipulation, and it's a little hard to follow. A few thoughts:

### Separating input and output

From reading through the code it seems like the intention is for the user to (rarely) use the routines in `preprocess/data_collection.py` to turn an event-name-to-metadata mapping (including paths to sample-containing HDF5 files) into a single `az.InferenceData` object that's saved to file. That file is later read in for analyses, and the routine for reading it and returning the dictionaries we typically use during inference is in `pipeline/utils.py`. Is that right? Does the library just use `az.InferenceData` objects because they're convenient for reading/writing, or do we use them elsewhere? Should we consider using them more extensively (e.g., providing them to the likelihood instead of dictionaries)?

It's generally good practice to have dedicated functions for reading and writing; e.g., [`save_posterior_samples_and_injections()` in `gwinferno/preprocess/data_collection.py`](https://github.com/FarrOutLab/GWInferno/blob/9dfdd57c955fbea2b0e65f2f36602c970c117735/gwinferno/preprocess/data_collection.py#L194) both uses a `json` file to find and load PE samples, construct an `InferenceData` object, and save that to file with a prescriptive file path.

### More descriptive function names

Separate from the choices we make above, it would be helpful to have more descriptive funcion names. Taking `data_collection.py` as an example, if I'm understanding things properly I think the following renames and modifications would make the code easier to follow:

- `run_map` > `catalog_metadata`
  
- `collect_data(run_map)` >`catalog_dict_from_metadata(catalog_metadata)`
  
- `format_data(posteriors)` > `catalog_xarray_from_dict(catalog_dict)`
  
- `evaluate_prior(full_catalog, param_names)` > `append_prior_to_catalog(catalog_xarray, param_names)`
  
- `save_posterior_samples_and_injections(key_file, injfile, parameter_names, outdir...)` > `inf_data_from_metadata_and_injections(catalog_metadata, injfile, parameter_names, ...)` where it returns the inference data object but doesn't write anthing to file.
  
- Then the user calls the `.to_netcdf()` method on the returned object, or we create a minimal wrapper function to do it, e.g., `save_catalog_inf_data(filename, catalog_inf_data)`. I think the only reason to have a function that forces a formatting to the filename is if we intend to have meaningful metadata as part of the filename, which doesn't seem necessary.

**Comment by jaxengodfrey** [comment]
> There's a _lot_ of data manipulation, and it's a little hard to follow. A few thoughts:
> 
> ### Separating input and output
> From reading through the code it seems like the intention is for the user to (rarely) use the routines in `preprocess/data_collection.py` to turn an event-name-to-metadata mapping (including paths to sample-containing HDF5 files) into a single `az.InferenceData` object that's saved to file. 

My intention is for the user to run `preprocess.data_collection.save_posterior_samples_and_injections` once, prior to running any analysis script, so that the processed data is saved in a format that is easily read in later with a simple `az.from_netcdf()` call. I do provide a function for reading in the idata object and converting it to a dictionary, but in theory the user could decide not to use this function, which I'll explain below. 

> That file is later read in for analyses, and the routine for reading it and returning the dictionaries we typically use during inference is in `pipeline/utils.py`. Is that right? 

I placed the `load_pe_and_injections_as_dict` function in `pipeline.utils` because I wanted to make it clear it was a function you should call in an analysis script. After the user has saved the processed PE and injections in an idata object, they should only ever need to use `load_pe_and_injections_as_dict` to then access that data. 

> Does the library just use `az.InferenceData` objects because they're convenient for reading/writing, or do we use them elsewhere? 

The short answer is yes, we use arviz and xarray because they are convenient to save/load data. I specifically need to use `az.InferenceData` because it lets me save multiple `xarray.Datasets` in one object/file, i.e. the PE Dataset and the injection Dataset. 

> Should we consider using them more extensively (e.g., providing them to the likelihood instead of dictionaries)?

The short answer is no, I don't think we want to do that. 

Here are my longer answers for why I chose to save/process the data with Arviz and Xarray while using the data as dictionaries in analysis scripts:

1. Dictionaries are easy to use and most users will know how to access the data stored in them. Initially, we used deepdish because it was very easy to save dictionaries to file. I haven't found a package like deepdish that saves dictionaries straight to file so I needed to find an alternative. 

2. h5py is an option, but in my opinion, saving and loading data with h5py is annoyingly cumbersome. While most users will likely not have any experience with xarrays, constructing xarrays is more intuitive to me and offers the benefit that once you've constructed them, you can use a simple .to_netcdf() call to save them to file. 

3. Xarray Datasets are great for organizing data and storing metadata, but not so great when you want to access and perform calculations on the individual DataArrays stored within them. For example, in the object saved by `save_posterior_samples_and_injections`, if I wanted to access the `mass_1` DataArray, I'd have to use:

```python
import arviz as az
idata = az.from_netcdf('data_file.h5') # idata is an InferenceData object with xarray Datasets pe_data and inj_data.

mass_1 = idata.pe_data.posteriors.sel(param='mass_1') # mass_1 is an xarray DataArray
```

 And if I wanted to then perform any sort of calculation on `mass_1` with `jax`, I'd have to add `.values` like in `jnp.exp(mass_1.values)`, because  'xarray.core.dataarray.DataArray' is not a valid `jax` type. Therefore, having the data readily accessible as a `jax.numpy.array` is the easiest way to perform calculations, which is the benefit of working with dictionaries. 

4. Lastly, the main piece of code the user will write on their own is their numpyro model.  `pipeline.analysis.hierarchical_likelihood` expects all array-like arguments to be `jax.numpy.arrays`. The user could decide to not use `pipeline.utils.load_pe_and_injections_as_dict` and instead load the data with `az.from_netcdf()` and work in xarrays until they have to plug in their arguments (like the weights) as `jax.numpy.arrays`. Or, they can choose to use `pipeline.utils.load_pe_and_injections_as_dict` and work in `jax.numpy.arrays` from the beginning. 

> It's generally good practice to have dedicated functions for reading and writing; e.g., [`save_posterior_samples_and_injections()` in `gwinferno/preprocess/data_collection.py`](https://github.com/FarrOutLab/GWInferno/blob/9dfdd57c955fbea2b0e65f2f36602c970c117735/gwinferno/preprocess/data_collection.py#L194) both uses a `json` file to find and load PE samples, construct an `InferenceData` object, and save that to file with a prescriptive file path.
> 
> ### More descriptive function names
> Separate from the choices we make above, it would be helpful to have more descriptive funcion names. Taking `data_collection.py` as an example, if I'm understanding things properly I think the following renames and modifications would make the code easier to follow:
> 
> * `run_map` > `catalog_metadata`
> * `collect_data(run_map)` >`catalog_dict_from_metadata(catalog_metadata)`
> * `format_data(posteriors)` > `catalog_xarray_from_dict(catalog_dict)`
> * `evaluate_prior(full_catalog, param_names)` > `append_prior_to_catalog(catalog_xarray, param_names)`
> * `save_posterior_samples_and_injections(key_file, injfile, parameter_names, outdir...)` 

I agree mostly with these name changes, though I think to make things a little more clear I would suggest these names:

* `collect_data(run_map)` > `raw_catalog_dict_from_metadata(catalog_metadata)`
* `format_data(posteriors)` > `processed_catalog_xarray_from_dict(catalog_dict)`
* `evaluate_prior(full_catalog, param_names)` > `append_prior_to_processed_catalog(catalog_xarray, param_names)`

> * `inf_data_from_metadata_and_injections(catalog_metadata, injfile, parameter_names, ...)` where it returns the inference data object but doesn't write anthing to file.
> * Then the user calls the `.to_netcdf()` method on the returned object, or we create a minimal wrapper function to do it, e.g., `save_catalog_inf_data(filename, catalog_inf_data)`. I think the only reason to have a function that forces a formatting to the filename is if we intend to have meaningful metadata as part of the filename, which doesn't seem necessary.

I see what you're saying, though we already have the functions `load_posterior_data` and `load_injections` that load the data without saving. The current function `save_posterior_samples_and_injections` just calls these functions internally, but we could make them arguments instead. What if we did this instead:

* the user loads the processed PE data as an `xarray.Dataset` with `load_posterior_data(catalog_metadata, param_names)`
* they load the processed injection data also as an `xarray.Dataset` with `load_injections(injfile, param_names, ...)`
* they then save these two Datasets to one file with `save_xarray_posterior_samples_and_injections_as_idata(filename, posteriors, injections). 

I like this because it reduces the number of arguments needed in a function like `inf_data_from_metadata_and_injections`. And if the user only wants to process one of the datasets, like the PE data, they don't have to also process the other dataset. 

## Final Thoughts

Hopefully this clears up the confusion. I think my example notebook on git.ligo.org is helpful for walking the user through these steps. I have a git issue for myself to make a new example script that goes through all the steps of using GWInferno. 

Ultimately, I want to make the data processing and loading as easy as possible, which means requiring the user to call as few functions as possible. I don't intend the average user to use the data manipulation functions in `data_collection` other than the three I mentioned above. Maybe we should put them in a different file like `data_collection_utils`. What do you think? I'll make the changes I mentioned and push to this branch so you can see what I'm thinking.

**Comment by bfarr** [comment]
Cool, I think this all makes sense.  While we could tweak these things endlessly, I think this is a huge improvement overall and ready to merge!

**Review by bfarr — APPROVED** [review]

**Comment by bfarr** [comment]
It looks like the unit tests just need to be adjusted, and *then* it's good to merge.
